[CSS Nesting] Don't serialize the implicit parent selector in relative selector

mdubet opened the pull request:
#### 22a69c73dd84b61a4a15ab5294b3880af1cb14a9
<pre>
[CSS Nesting] Don&apos;t serialize the implicit parent selector in relative selector
<a href="https://bugs.webkit.org/show_bug.cgi?id=256578">https://bugs.webkit.org/show_bug.cgi?id=256578</a>
rdar://109141615

Reviewed by Antti Koivisto.

Instead of inserting the implicit parent selector at parse time
(which makes it appears in the serialization),
we insert the resolved parent later (during rule set building).

* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/implicit-nesting-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/implicit-nesting-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/implicit-nesting.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/parsing.html:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::selectorText const):
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeStyleRule):
(WebCore::CSSParserImpl::consumeDeclarationListOrStyleBlockHelper):
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::parseCSSSelector):
(WebCore::CSSSelectorParser::consumeRelativeNestedSelector):
(WebCore::CSSSelectorParser::resolveNestingParent):

Canonical link: <a href="https://commits.webkit.org/263902@main">https://commits.webkit.org/263902@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/494df1db468f64cc883471de14392b3a17b365d8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6075 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6257 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6441 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7635 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6422 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6074 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6475 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6211 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9299 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6184 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6205 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5495 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7697 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3680 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5476 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13386 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5546 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5554 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/7786 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6020 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4925 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5443 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/5405 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1436 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9581 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5810 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -861,6 +874,9 @@ String CSSSelector::selectorText(StringView separator, StringView rightSide) con
         return previousSelector->selectorText(separator, builder);
+    } else if (!separatorTextForNestingRelative().isNull()) {
+        // We have a separator but no tag history which can happen with implicit relative nesting selector
+        return separatorTextForNestingRelative() + builder.toString();
```
it would be more stylish to not call separatorTextForNestingRelative() twice, for example.
```
} else if (auto separatorText = separatorTextForNestingRelative(); !separatorText.isNull()) 
    return separatorText + builder.toString();
```

mdubet committed:
[CSS Nesting] Don't serialize the implicit parent selector in relative selector
https://bugs.webkit.org/show_bug.cgi?id=256578
rdar://109141615

Reviewed by Antti Koivisto.

Instead of inserting the implicit parent selector at parse time
(which makes it appears in the serialization),
we insert the resolved parent later (during rule set building).

* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/implicit-nesting-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/implicit-nesting-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/implicit-nesting.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/parsing.html:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::selectorText const):
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeStyleRule):
(WebCore::CSSParserImpl::consumeDeclarationListOrStyleBlockHelper):
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::parseCSSSelector):
(WebCore::CSSSelectorParser::consumeRelativeNestedSelector):
(WebCore::CSSSelectorParser::resolveNestingParent):

Canonical link: https://commits.webkit.org/263902@main
